src/search.c

```diff
@@ -178,10 +178,21 @@ rs_search_for_block(rs_weak_sum_t weak_sum,
     int l = bucket->l; /* left bound of search region */
     int r = bucket->r + 1; /* right bound of search region */
     int v = 1; /* direction of next move: -ve left, +ve right */
+    int i = -1;
 
     if (l == NULL_TAG)
         return 0;
 
+    /* Try a direct match */
+    if ((job_token > 0) && (job_token < sig->count) &&
+        (weak_sum == sig->block_sigs[job_token-1].weak_sum)) {
```

Comment:
Maybe, but testing first if the same block in the source file matches seems to be a good guess. If a file has multiple identical blocks, which one do you take as a reference? The first one, the last one...depends on the implementation of the hash table. Since they are identical it doesn't matter where you take the data (it's the same), but delta file will be smaller if blocks can be merged together (so you get only a single operation spanning multiple blocks).